### PR TITLE
Add `uniq(list)`, `uniq(string)` helper function and update related tests

### PR DESCRIPTION
**Pull request by mjkim610**
## Proposed changes

This PR adds the `uniq(slice)` and `uniq(string)` helper functions as described in https://github.com/projectdiscovery/nuclei/issues/2305#issue-1307747022 and https://github.com/projectdiscovery/nuclei/issues/2305#issuecomment-1187426430

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mjkim610** [comment]
@forgedhallpass @Mzack9999 please review :)

File: v2/pkg/operators/common/dsl/dsl.go [hunk 1]
```diff
@@ -77,6 +77,37 @@ func init() {
...
 			return strings.ToLower(types.ToString(args[0])), nil
 		}),
+		"uniq": makeDslWithOptionalArgsFunction(
+			"(args ...interface{}) interface{}",
+			func(args ...interface{}) (interface{}, error) {
+				argCount := len(args)
+				if argCount <= 0 {
+					return nil, invalidDslFunctionError
+				} else if argCount == 1 {
+					result := ""
```
Comment (Mzack9999) [inline v2/pkg/operators/common/dsl/dsl.go:87]: `strings.Builder` is more efficient
```go
var result strings.Builder
...
builder.WriteRune(i) // i being a rune variable
```

File: v2/pkg/operators/common/dsl/dsl.go [hunk 2]
```diff
@@ -77,6 +77,37 @@ func init() {
...
+					visited := make(map[rune]bool)
```
Comment (Mzack9999) [inline v2/pkg/operators/common/dsl/dsl.go:88]: `bool` => `struct{}` (uses less memory)
```go
visited := make(map[rune]struct{})
```

File: v2/pkg/operators/common/dsl/dsl.go [hunk 3]
```diff
@@ -77,6 +77,37 @@ func init() {
...
+						_, isRuneSeen := visited[i]
+						if !isRuneSeen {
+							result += string(i)
+							visited[i] = true
+						}
+					}
+					return result, nil
+				} else {
+					result := make([]string, 0, argCount)
+					visited := make(map[string]bool)
```
Comment (Mzack9999) [inline v2/pkg/operators/common/dsl/dsl.go:99]: `bool` => `struct{}`

File: v2/pkg/operators/common/dsl/dsl.go [hunk 4]
```diff
@@ -339,6 +370,24 @@ func init() {
...
+		"join_slice": makeDslWithOptionalArgsFunction(
+			"(separator string, elements ...interface{}) string",
+			func(arguments ...interface{}) (interface{}, error) {
+				argumentsSize := len(arguments)
+				if argumentsSize != 2 {
+					return nil, errors.New("incorrect number of arguments received")
+				}
+
+				separator := types.ToString(arguments[0])
+				elements := arguments[1].([]string)
```
Comment (Mzack9999) [inline v2/pkg/operators/common/dsl/dsl.go:382]: the cast should be checked here:
```
if elements, ok := arguments[1].([]string); ok {
```

File: v2/pkg/operators/common/dsl/dsl.go [hunk 5]
```diff
@@ -339,6 +370,24 @@ func init() {
...
+
+				stringElements := make([]string, 0, len(elements))
+				for _, element := range elements {
+					stringElements = append(stringElements, types.ToString(element))
+				}
+				return strings.Join(stringElements, separator), nil
```
Comment (Mzack9999) [inline v2/pkg/operators/common/dsl/dsl.go:388]: `elements` is already `[]string`, you can use it directly instead of `stringElements`

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: v2/pkg/operators/common/dsl/dsl.go (hunk 2 above)
Comment (mjkim610) [reply v2/pkg/operators/common/dsl/dsl.go]: Done.

File: v2/pkg/operators/common/dsl/dsl.go (hunk 1 above)
Comment (mjkim610) [reply v2/pkg/operators/common/dsl/dsl.go]: Done.

File: v2/pkg/operators/common/dsl/dsl.go (hunk 3 above)
Comment (mjkim610) [reply v2/pkg/operators/common/dsl/dsl.go]: Done.

File: v2/pkg/operators/common/dsl/dsl.go (hunk 4 above)
Comment (mjkim610) [reply v2/pkg/operators/common/dsl/dsl.go]: Done.

File: v2/pkg/operators/common/dsl/dsl.go (hunk 5 above)
Comment (mjkim610) [reply v2/pkg/operators/common/dsl/dsl.go]: Done.

**Review by Mzack9999 — APPROVED** [review]

**Review by forgedhallpass — CHANGES_REQUESTED** [review]
The same comment as in the other PR. Please remove the `join_slice` from this PR and then we can merge it.

https://github.com/projectdiscovery/nuclei/pull/2372#discussion_r961666373

**Comment by mjkim610** [comment]
> The same comment as in the other PR. Please remove the `join_slice` from this PR and then we can merge it.
> 
> [#2372 (comment)](https://github.com/projectdiscovery/nuclei/pull/2372#discussion_r961666373)

Hi @forgedhallpass, thanks for the feedback :)

Do you want to remove the logic in `join_slice()` completely, or do you want to integrate its logic into `join()`?

**Comment by forgedhallpass** [comment]
The `join_slice` method should be fully integrated into the `join` method. Just make sure that you handle the corner cases as well (e.g. user passing in a variable number of slices. You either join all of them, or throw an exception saying that in case of slices, only one is expected). You should also make sure the changes are present in only one of your PRs, otherwise there'll be merge conflicts.

**Comment by mjkim610** [comment]
Changes in this PR is included in https://github.com/projectdiscovery/nuclei/pull/2372

**Pull request closed without merging**